<!--
  1 failure:  `danger-swift` fa...
  0 warning: 
  
  1 markdown notices
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  `danger-swift` failed.
  </td>
    </tr>
  </tbody>
</table>



### Log

<details>


```
/var/folders/b8/9zt_tx3s1y37nzh_cs991xyc0000gn/T/danger/ABC4EF93-65E7-45F7-9350-DEA624D99C7E/_tmp_dangerfile.swift:9:14: error: cannot call value of non-function type 'module<Danger>'
 7 | import Foundation
 8 | 
 9 | let danger = Danger()
   |              `- error: cannot call value of non-function type 'module<Danger>'
10 | 
11 | lint()

/var/folders/b8/9zt_tx3s1y37nzh_cs991xyc0000gn/T/danger/ABC4EF93-65E7-45F7-9350-DEA624D99C7E/_tmp_dangerfile.swift:17:5: error: cannot find 'message' in scope
15 | 
16 | func changedFiles() {
17 |     message("Edited \(danger.git.modifiedFiles.count) files")
   |     `- error: cannot find 'message' in scope
18 |     message("Created \(danger.git.createdFiles.count) files")
19 | }

/var/folders/b8/9zt_tx3s1y37nzh_cs991xyc0000gn/T/danger/ABC4EF93-65E7-45F7-9350-DEA624D99C7E/_tmp_dangerfile.swift:18:5: error: cannot find 'message' in scope
16 | func changedFiles() {
17 |     message("Edited \(danger.git.modifiedFiles.count) files")
18 |     message("Created \(danger.git.createdFiles.count) files")
   |     `- error: cannot find 'message' in scope
19 | }
20 | 

/var/folders/b8/9zt_tx3s1y37nzh_cs991xyc0000gn/T/danger/ABC4EF93-65E7-45F7-9350-DEA624D99C7E/_tmp_dangerfile.swift:22:22: error: cannot find 'SwiftLint' in scope
20 | 
21 | func lint() {
22 |     let violations = SwiftLint.lint(.all(directory: "./"), inline: true, configFile: ".swiftlint.yml", strict: true, quiet: false)
   |                      `- error: cannot find 'SwiftLint' in scope
23 |     if violations.isEmpty {
24 |         message("No SwiftLint violations! 🎉")

/var/folders/b8/9zt_tx3s1y37nzh_cs991xyc0000gn/T/danger/ABC4EF93-65E7-45F7-9350-DEA624D99C7E/_tmp_dangerfile.swift:22:38: error: cannot infer contextual base in reference to member 'all'
20 | 
21 | func lint() {
22 |     let violations = SwiftLint.lint(.all(directory: "./"), inline: true, configFile: ".swiftlint.yml", strict: true, quiet: false)
   |                                      `- error: cannot infer contextual base in reference to member 'all'
23 |     if violations.isEmpty {
24 |         message("No SwiftLint violations! 🎉")

/var/folders/b8/9zt_tx3s1y37nzh_cs991xyc0000gn/T/danger/ABC4EF93-65E7-45F7-9350-DEA624D99C7E/_tmp_dangerfile.swift:24:9: error: cannot find 'message' in scope
22 |     let violations = SwiftLint.lint(.all(directory: "./"), inline: true, configFile: ".swiftlint.yml", strict: true, quiet: false)
23 |     if violations.isEmpty {
24 |         message("No SwiftLint violations! 🎉")
   |         `- error: cannot find 'message' in scope
25 |     }
26 | }

/var/folders/b8/9zt_tx3s1y37nzh_cs991xyc0000gn/T/danger/ABC4EF93-65E7-45F7-9350-DEA624D99C7E/_tmp_dangerfile.swift:30:9: error: cannot find 'fail' in scope
28 | func coverage() {
29 |     guard let xcresult = ProcessInfo.processInfo.environment["BITRISE_XCRESULT_PATH"]?.escapeString() else {
30 |         fail("Could not get the BITRISE_XCRESULT_PATH to generage code coverage")
   |         `- error: cannot find 'fail' in scope
31 |         return
32 |     }

/var/folders/b8/9zt_tx3s1y37nzh_cs991xyc0000gn/T/danger/ABC4EF93-65E7-45F7-9350-DEA624D99C7E/_tmp_dangerfile.swift:34:5: error: cannot find 'Coverage' in scope
32 |     }
33 | 
34 |     Coverage.xcodeBuildCoverage(
   |     `- error: cannot find 'Coverage' in scope
35 |         .xcresultBundle(xcresult),
36 |         minimumCoverage: 50

/var/folders/b8/9zt_tx3s1y37nzh_cs991xyc0000gn/T/danger/ABC4EF93-65E7-45F7-9350-DEA624D99C7E/_tmp_dangerfile.swift:35:10: error: cannot infer contextual base in reference to member 'xcresultBundle'
33 | 
34 |     Coverage.xcodeBuildCoverage(
35 |         .xcresultBundle(xcresult),
   |          `- error: cannot infer contextual base in reference to member 'xcresultBundle'
36 |         minimumCoverage: 50
37 |     )

/var/folders/b8/9zt_tx3s1y37nzh_cs991xyc0000gn/T/danger/ABC4EF93-65E7-45F7-9350-DEA624D99C7E/_tmp_dangerfile.swift:41:5: error: cannot find 'message' in scope
39 | 
40 | func xcodeEnvironmentVariables() {
41 |     message("Checking XCode Environment Variables")
   |     `- error: cannot find 'message' in scope
42 | 
43 |     let snowplowPostPath = #"""
[31mERROR: Dangerfile eval failed at Dangerfile.swift
[0;0m[31mERROR: Could not get the results JSON file at /var/folders/b8/9zt_tx3s1y37nzh_cs991xyc0000gn/T/danger/ABC4EF93-65E7-45F7-9350-DEA624D99C7E/danger-response.json
[0;0m
```


</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 0a2464239b99b97b1ed5df4c7fb702cc5d64b79d
</p>

--- a/PocketKit/Sources/PocketKit/Home/SwiftUI/Models/HomeCardModel.swift
+++ b/PocketKit/Sources/PocketKit/Home/SwiftUI/Models/HomeCardModel.swift
@@ -6,15 +6,20 @@ import Localization
 import Sync
 import SwiftUI
 
+// TODO: SWIFTUI - Add analytics
+
+@MainActor
 @Observable
 final class HomeCardModel {
     let item: Item
     let overflowActions: [HomeButtonAction]
-    let primaryAction: HomeButtonAction?
+    private(set) var primaryAction: HomeButtonAction?
     let favoriteAction: HomeButtonAction?
     var imageURL: URL?
     var title: String?
     var sharedWithYouUrlString: String?
+
+    private let source = Services.shared.source
 
     var isSaved: Bool {
         item.savedItem != nil &&
@@ -32,11 +37,19 @@ final class HomeCardModel {
     ) {
         self.item = item
         self.overflowActions = overflowActions
-        self.primaryAction = primaryAction
         self.favoriteAction = favoriteAction
         self.imageURL = imageURL
         self.title = title ?? item.syndicatedArticle?.title ?? item.title
         self.sharedWithYouUrlString = sharedWithYouUrlString
+
+        self.primaryAction = HomeButtonAction { [weak self] in
+            guard let self else { return }
+            if isSaved {
+                source.archive(from: item.givenURL)
+            } else {
+                source.save(from: item.givenURL)
+            }
+        }
     }
 }
 

--- a/PocketKit/Sources/PocketKit/Home/SwiftUI/Models/HomeCardModel.swift
+++ b/PocketKit/Sources/PocketKit/Home/SwiftUI/Models/HomeCardModel.swift
@@ -18,7 +18,7 @@ final class HomeCardModel {
     var imageURL: URL?
     var title: String?
     var sharedWithYouUrlString: String?
-
+    // TODO: SWIFTUI - Once we are fully migrated to SwiftUI, this should come from the environment.
     private let source = Services.shared.source
 
     var isSaved: Bool {

--- a/PocketKit/Sources/PocketKit/Home/SwiftUI/Views/HomeCarouselView.swift
+++ b/PocketKit/Sources/PocketKit/Home/SwiftUI/Views/HomeCarouselView.swift
@@ -3,17 +3,26 @@
 // file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
 import Localization
+import SwiftData
 import SwiftUI
+import Sync
 import Textile
 
 struct HomeCarouselView: View {
     var model: HomeCardModel
-    // TODO: this will reflect the state of the saved item
-    @State private var isSaved: Bool = false
-    @State private var isFavorited: Bool = false
+
+    @Query var savedItem: [SavedItem]
+    var currentSavedItem: SavedItem? {
+        savedItem.first
+    }
 
     init(model: HomeCardModel) {
         self.model = model
+
+        let givenUrl = model.item.givenURL
+        var descriptor = FetchDescriptor<SavedItem>(predicate: #Predicate<SavedItem> { $0.item?.givenURL == givenUrl })
+        descriptor.fetchLimit = 1
+        _savedItem = Query(descriptor, animation: .default)
     }
 
     var body: some View {
@@ -99,7 +108,7 @@ private extension HomeCarouselView {
 
     func makeFavoriteButton(handler: @escaping (() -> Void)) -> some View {
         HomeActionButton(
-            isActive: isFavorited,
+            isActive: currentSavedItem?.isFavorite == false,
             activeImage: .favoriteFilled,
             inactiveImage: .favorite,
             highlightedColor: .branding.amber1,
@@ -113,7 +122,7 @@ private extension HomeCarouselView {
 
     func makeSaveButton(handler: @escaping (() -> Void)) -> some View {
         HomeActionButton(
-            isActive: isSaved,
+            isActive: currentSavedItem?.isArchived == false,
             activeImage: .saved,
             inactiveImage: .save,
             activeTitle: Localization.Recommendation.saved,
@@ -121,7 +130,7 @@ private extension HomeCarouselView {
             highlightedColor: .ui.coral1,
             activeColor: .ui.coral2
         ) {
-            handler()
+            model.primaryAction?.action()
         }
         .accessibilityIdentifier("save-button")
     }

--- a/PocketKit/Sources/PocketKit/Home/SwiftUI/Views/HomeHeroView.swift
+++ b/PocketKit/Sources/PocketKit/Home/SwiftUI/Views/HomeHeroView.swift
@@ -11,8 +11,8 @@ import Textile
 
 struct HomeHeroView: View {
     var model: HomeCardModel
-    @Query var savedItem: [SavedItem]
 
+    @Query var savedItem: [SavedItem]
     var currentSavedItem: SavedItem? {
         savedItem.first
     }
@@ -108,7 +108,7 @@ private extension HomeHeroView {
             highlightedColor: .ui.coral1,
             activeColor: .ui.coral2
         ) {
-            // isSaved.toggle()
+            model.primaryAction?.action()
         }
         .accessibilityIdentifier("save-button")
     }

--- a/PocketKit/Sources/Sync/PocketSource.swift
+++ b/PocketKit/Sources/Sync/PocketSource.swift
@@ -1212,6 +1212,23 @@ extension PocketSource {
         }
     }
 
+    public func save(from givenURL: String) {
+        space.performAndWait {
+            guard let item = try? space.fetchItem(byURL: givenURL) else {
+                return
+            }
+            if let savedItem = try? space.fetchSavedItem(byURL: givenURL) {
+                unarchive(item: savedItem)
+            } else {
+                let savedItem: CDSavedItem = CDSavedItem(context: space.backgroundContext, url: givenURL)
+                savedItem.update(from: item)
+                try? space.save()
+
+                save(item: savedItem)
+            }
+        }
+    }
+
     public func save(item: CDItem) {
         space.performAndWait {
             guard let item = space.backgroundObject(with: item.objectID) as? CDItem else {
@@ -1264,6 +1281,15 @@ extension PocketSource {
         space.performAndWait {
             guard let collectionStory = space.backgroundObject(with: collectionStory.objectID) as? CDCollectionStory,
                   let savedItem = collectionStory.item?.savedItem, savedItem.isArchived == false else {
+                return
+            }
+            archive(item: savedItem)
+        }
+    }
+
+    public func archive(from givenURL: String) {
+        space.performAndWait {
+            guard let item = try? space.fetchItem(byURL: givenURL), let savedItem = item.savedItem else {
                 return
             }
             archive(item: savedItem)

--- a/PocketKit/Sources/Sync/Source.swift
+++ b/PocketKit/Sources/Sync/Source.swift
@@ -84,11 +84,15 @@ public protocol Source {
 
     func save(recommendation: CDRecommendation)
 
+    func save(from givenURL: String)
+
     func save(item: CDItem)
 
     func save(collectionStory: CDCollectionStory)
 
     func archive(recommendation: CDRecommendation)
+
+    func archive(from givenURL: String)
 
     func archive(collectionStory: CDCollectionStory)
 

--- a/PocketKit/Tests/PocketKitTests/Support/MockSource.swift
+++ b/PocketKit/Tests/PocketKitTests/Support/MockSource.swift
@@ -8,6 +8,10 @@ import CoreData
 import Combine
 
 class MockSource: Source {
+    func archive(from givenURL: String) {}
+
+    func save(from givenURL: String) {}
+
     func readerItem(by slug: String) async throws -> (Sync.CDSavedItem?, Sync.CDItem?) {
         return (nil, nil)
     }

